core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -438,18 +438,33 @@ public boolean hasAliases(final String[] aliases, String[] concreteIndices) {
 
     /**
      * Returns indexing routing for the given index.
+     * If routing is not null then the value is returned, parent routing and alias routing are ignored.
+     * If routing is null and there is an alias routing then the value is returned and parent is ignored.
+     * If routing is null and a parent is defined then the parent is returned (unless an alias routing is defined, see above).
      */
     // TODO: This can be moved to IndexNameExpressionResolver too, but this means that we will support wildcards and other expressions
     // in the index,bulk,update and delete apis.
-    public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex) {
-        if (aliasOrIndex == null) {
+    public String resolveIndexRouting(@Nullable String parent, @Nullable String routing, String aliasOrIndex) {
+        if (routing != null) {
+            // The routing is explicit, ignore the alias and parent routing (if any).
```

Comment:
Can we not ignore but throw an error back to the user? We should silently ignore something the user has passed in, it may be they are confused about the api and don't realize that part of their request is being ignored.